### PR TITLE
Jwt changes for issue #2100

### PR DESCRIPTION
**Pull request by amit2103**
Only java code changes, will submit the JS changes by this weekend.

### CODE REVIEW

**Comment by deepu105** [comment]
@amit2103 can you take look at the conflicts and resolve it plz?

**Comment by jdubois** [comment]
I have an issue with this PR: the idea with JHipster is to use Spring Security, which provides "spring-security-jwt".
I've never used it or even tried it, but normally this is the library we should rely on. Besides, here there is quite a lot of code, and I'd rather have something that is already built-in Spring Security.
Can we have at least some discussion on why this is not being used here?

**Comment by amit2103** [comment]
@deepu105 I will do that.
@jdubois The Spring Security JWT project is not maintained well. Moving to that is not a major change, 
however I think keeping the JWT token creation easy and configurable will be better. Not much of the code will change except the Token creation filters. The rest remains the same. Let me know what you think.

**Comment by amit2103** [comment]
Guys If you think we should implement the Spring Security standard , I can do that too. I think that Spring Security project is not maintained well. However you can take a look and decide.

**Comment by deepu105** [comment]
Amit im not very familiar with jwt, so lets wait for @jdubois to comment

**Comment by amit2103** [comment]
@deepu105  yea..I understand...

**Comment by pnreddysvu** [comment]
yes spring-jwt is not at all maintained. After doing lot of study I ended up choosing https://github.com/jwtk/jjwt which is most promising at this point.

**Comment by gmarziou** [comment]
Do you have any evidence that spring-security-jwt is not well maintained? 

I just looked at [opened issues}(https://github.com/spring-projects/spring-security-oauth/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+jwt) and there are not many and even the oldest ones have active and recent discussions.

**Comment by pnreddysvu** [comment]
@gmarziou Below are my points when I say it is not well maintained.
1) Spring-jwt can only be used along with OAuth2.
2) Spring-jwt does not implement all specs ex: JWE
3) Spring-jwt does not have support for optional fields defined in spec. Ex: Issuer claims

Considering all of those we decided to go with 3rd party OSS JWT implementations for ourproject. We shortlisted 2 such implementations.

http://connect2id.com/products/nimbus-jose-jwt
https://github.com/jwtk/jjwt

Both are really good. We used jjwt.

**Comment by pnreddysvu** [comment]
@amit2103  You should use Authorization header instead of 'X-AUTH-TOKEN' header.

**Comment by amit2103** [comment]
@pnreddysvu Yes, I choose JJWT pretty much for the same reason.
Yes I can use the Authorization header. Lets first decide whether we want to stick to jjwt or Spring Security. I will make the requisite changes then.

Also check this https://github.com/spring-projects/spring-security-oauth/tree/master/spring-security-jwt.

The last commit is 9 months ago. This might indicate a stable product or inactivity. I think its the later. All the examples and blogs talk of avoiding this due to similar reasons. I also think that jjwt api is cleaner and easier to understand.

**Comment by gmarziou** [comment]
> 1) Spring-jwt can only be used along with OAuth2.

Well, it has its own pom.xml and is released independently on maven central. What do you mean?

> I also think that jjwt api is cleaner and easier to understand.

In [this issue](https://github.com/spring-projects/spring-security-oauth/issues/575#issuecomment-146858378) someone compared jjwt and nimbus.

**Comment by pnreddysvu** [comment]
See the issue https://github.com/spring-projects/spring-security-oauth/issues/368 for more details regarding spring-jwt.

The comparision was not between Nimbus and jjwt it was between Nibus and jose4j which is a different library

**Comment by gmarziou** [comment]
Thanks for the clarifications, ou convinced me.

**Comment by deepu105** [comment]
@jdubois I think @amit2103 is waiting for confirmation from you on use of sprint jwt vs jjwt

File: app/index.js [hunk 1]
```diff
@@ -771,12 +769,12 @@ JhipsterGenerator.prototype.app = function app() {
     this.template('src/main/java/package/security/_UserDetailsService.java', javaDir + 'security/UserDetailsService.java', this, {});
     this.template('src/main/java/package/security/_UserNotActivatedException.java', javaDir + 'security/UserNotActivatedException.java', this, {});
 
-    if (this.authenticationType == 'xauth') {
-        this.template('src/main/java/package/security/xauth/_Token.java', javaDir + 'security/xauth/Token.java', this, {});
-        this.template('src/main/java/package/security/xauth/_TokenProvider.java', javaDir + 'security/xauth/TokenProvider.java', this, {});
-        this.template('src/main/java/package/web/rest/_UserXAuthTokenController.java', javaDir + 'web/rest/UserXAuthTokenController.java', this, {});
-        this.template('src/main/java/package/security/xauth/_XAuthTokenConfigurer.java', javaDir + 'security/xauth/XAuthTokenConfigurer.java', this, {});
-        this.template('src/main/java/package/security/xauth/_XAuthTokenFilter.java', javaDir + 'security/xauth/XAuthTokenFilter.java', this, {});
+     if (this.authenticationType == 'jwt') {
```
Comment (deepu105) [inline app/index.js]: extra space

File: app/templates/_pom.xml [hunk 2]
```diff
@@ -495,7 +495,17 @@
             <groupId>org.springframework.security.oauth</groupId>
             <artifactId>spring-security-oauth2</artifactId>
             <version>${spring-security-oauth2.version}</version>
-        </dependency>
+        </dependency>   
+        <%_ } _%>
+
```
Comment (deepu105) [inline app/templates/_pom.xml]: should there be empty line here?

File: app/templates/src/main/java/package/config/_SecurityConfiguration.java [hunk 3]
```diff
@@ -165,10 +174,10 @@ protected void configure(HttpSecurity http) throws Exception {
     @Bean
     public AuthenticationManager authenticationManagerBean() throws Exception {
         return super.authenticationManagerBean();
-    }<% } %><% if (authenticationType == 'xauth') { %>
+    }<% } %><% if (authenticationType == 'jwt') { %>
 
-    private XAuthTokenConfigurer securityConfigurerAdapter() {
-      return new XAuthTokenConfigurer(userDetailsService, tokenProvider);
+    private JWTTokenConfigurer securityConfigurerAdapter() {
+      return new JWTTokenConfigurer(userDetailsService, tokenAuthenticationService);
```
Comment (deepu105) [inline app/templates/src/main/java/package/config/_SecurityConfiguration.java]: it should be 4 space

File: app/templates/src/main/java/package/config/_SecurityConfiguration.java [hunk 4]
```diff
@@ -1,8 +1,8 @@
 package <%=packageName%>.config;
-<% if (authenticationType == 'session' || authenticationType == 'xauth') { %>
+<% if (authenticationType == 'session' || authenticationType == 'jwt') { %>
 import <%=packageName%>.security.*;<% } %><% if (authenticationType == 'session') { %>
-import <%=packageName%>.web.filter.CsrfCookieGeneratorFilter;<% } %><% if (authenticationType == 'xauth') { %>
-import <%=packageName%>.security.xauth.*;<% } %>
+import <%=packageName%>.web.filter.CsrfCookieGeneratorFilter;<% } %><% if (authenticationType == 'jwt') { %>
+import <%=packageName%>.security.jwt.*;<% } %>
```
Comment (deepu105) [inline app/templates/src/main/java/package/config/_SecurityConfiguration.java]: do we need * import as only one class seems to be used

**Comment by jdubois** [comment]
First of all I'd like to reach @lhazlewood who is the main commiter on jjwt:

@lhazlewood : 

- We are considering using jjwt for JHipster
- Usually that means we're going to send you quite a lot of users. Have a look at what happened to MapStruct when we start making them a (non-default!) option: https://twitter.com/gunnarmorling/status/618063720644784128
- Would you be OK with helping us support jjwt on JHipster? We are probably going to send you some issues (considering our number of users), so do you have the bandwidth to help us?

Then let's talk about our initial topic, on using jjwt. I'm not an expert in that field, all I see is that our current custom implementation should be removed to use a third-party library, as:

- It's not our job to implement such specifications
- We are not security experts, and those third-party libs will probably do a better job than what we could do

Then I always try to focus on the Spring libraries, as they provide a consistent programming model, but of course Spring can't do everything perfectly: for example I'm not using Spring Data Cassandra for our Cassandra support, because I don't find it good enough.

So that's probably the same thing here: from my understanding, jjwt looks to be a much better solution, so let's try to go that way, as long as the jjwt project is OK with helping us when we have some issues.

**Comment by amit2103** [comment]
@deepu105  Will fix the typos..Will take me till this weekend though. Quite busy at work.
@jdubois  I will work on the other review comments. Let me know if you have any feedback, I had started work on the client then stopped. I will start working on it again.

**Comment by lhazlewood** [comment]
@jdubois et. al:

Hi folks - I'm happy to support issues that arise from JHipster users.  Whether it is me or another [Stormpath](https://stormpath.com) Java team member, we have the bandwidth to help.  We are actively maintaining JJWT and improving it all the time.  My only request is that issues created for us are actually JJWT issues (as we wouldn't have knowledge enough to troubleshoot JHipster-related issues).

Happy to help!

**Comment by jdubois** [comment]
Thanks @lhazlewood !!
Let's go and move to jjwt then :-)

**Comment by deepu105** [comment]
@amit2103 can you rebase to master and let us know if this is ready?

**Comment by cicoub13** [comment]
I'm very interested in this feature. If i can help.

**Comment by jdubois** [comment]
@cicoub13 this looks stuck for a couple of months, can you have a look at the PR and maybe take the lead on this?
Please note that since this time, I have started to work on our microservices architecture, and I'm using this [tutorial](https://spring.io/blog/2015/02/03/sso-with-oauth2-angular-js-and-spring-security-part-v) as a base example, which uses Spring's JWT implementation. So maybe there is something to align here.

**Comment by amit2103** [comment]
@deepu105  Will do it this weekend..have completed the changes.

**Comment by amit2103** [comment]
@jdubois  Have completed the changes...not tested yet..will update this by this weekend..after the pull request you can review and then may be we can think of alignment.

**Comment by jdubois** [comment]
Thanks @amit2103 I'm very interested by your work

File: app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package <%=packageName%>.security.xauth;
```
Comment (jdubois) [inline app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java]: shouldn't the package name be .security.jwt ? or .security.jjwt?

File: app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.security.web.DefaultSecurityFilterChain;
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+
+public class JWTTokenConfigurer extends SecurityConfigurerAdapter<DefaultSecurityFilterChain, HttpSecurity> {
+
+    private TokenAuthenticationService tokenAuthenticationService;
+
+    private UserDetailsService detailsService;
+
+    public JWTTokenConfigurer(UserDetailsService detailsService, TokenAuthenticationService tokenAuthenticationService) {
```
Comment (jdubois) [inline app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java]: Isn't this a Spring Bean? I need to check but the dependency injection isn't set up

File: app/templates/src/main/java/package/security/jwt/_TokenAuthenticationService.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Created by jhipster on 10/25/15.
+ */
+@Service
+public class TokenAuthenticationService {
+
+    private static final String AUTH_HEADER_NAME = "X-AUTH-TOKEN";
```
Comment (jdubois) [inline app/templates/src/main/java/package/security/jwt/_TokenAuthenticationService.java]: We need to discuss this -> the "X-" should now not be used (latest RFC...) and as this is specific to JHipster I'd do something like "X-JHIPSTER-TOKEN" or "X-JHIPSTER-AUTH"

**Comment by jdubois** [comment]
I haven't tried this but this looks *very* promising with regards to https://github.com/jhipster/jhipster-experimental-microservices/issues/3 (cc @sdoxsee ) in a microservices architecture.
That would works with or without OAuth2 at the gateway level, and seems very easy to use.
This is on the top of my "must have" features.

File: app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java (hunk 5 above)
Comment (amit2103) [inline app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java]: Yes...Will rename this.

File: app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java (hunk 6 above)
Comment (amit2103) [inline app/templates/src/main/java/package/security/jwt/_JWTTokenConfigurer.java]: Yes...Will set this up

File: app/templates/src/main/java/package/security/jwt/_TokenAuthenticationService.java (hunk 7 above)
Comment (amit2103) [inline app/templates/src/main/java/package/security/jwt/_TokenAuthenticationService.java]: Second one looks better to me

**Comment by amit2103** [comment]
@jdubois  Thanks for the comments..Will update all the  java files rebased and updated with master by tomorrow.

**Comment by amit2103** [comment]
Created a new pull request this one seems to be very old.

**Comment by jdubois** [comment]
OK @amit2103 so I'm closing this because of #2746

**Pull request closed without merging**